vendor/k8s.io/kube-openapi/pkg/schemaconv/openapi.go

```diff
@@ -0,0 +1,257 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schemaconv
+
+import (
+	"errors"
+	"path"
+	"strings"
+
+	"k8s.io/kube-openapi/pkg/validation/spec"
+	"sigs.k8s.io/structured-merge-diff/v4/schema"
+)
+
+// ToSchemaFromOpenAPI converts a directory of OpenAPI schemas to an smd Schema.
+//   - models: a map from definition name to OpenAPI V3 structural schema for each definition.
+//     Key in map is used to resolve references in the schema.
+//   - preserveUnknownFields: flag indicating whether unknown fields in all schemas should be preserved.
+//   - returns: nil and an error if there is a parse error, or if schema does not satisfy a
+//     required structural schema invariant for conversion. If no error, returns
+//     a new smd schema.
+//
+// Schema should be validated as structural before using with this function, or
+// there may be information lost.
+func ToSchemaFromOpenAPI(models map[string]*spec.Schema, preserveUnknownFields bool) (*schema.Schema, error) {
+	c := convert{
+		preserveUnknownFields: preserveUnknownFields,
+		output:                &schema.Schema{},
+	}
+
+	for name, spec := range models {
+		// Skip/Ignore top-level references
+		if len(spec.Ref.String()) > 0 {
+			continue
+		}
+
+		var a schema.Atom
+		if name == quantityResource || name == rawExtensionResource {
+			//!TODO: add support to the openapi schema to properly support these
+			// Hardcode both quantity and raw extension to use untyped schemas
+			a = schema.Atom{
+				Scalar: ptr(schema.Scalar("untyped")),
+			}
+		} else {
+			c2 := c.push(name, &a)
+			c2.visitSpec(spec)
+			c.pop(c2)
+		}
+
+		c.insertTypeDef(name, a)
+	}
+
+	if len(c.errorMessages) > 0 {
+		return nil, errors.New(strings.Join(c.errorMessages, "\n"))
+	}
+
+	c.addCommonTypes()
+	return c.output, nil
+}
+
+func (c *convert) visitSpec(m *spec.Schema) {
+	// Check if this schema opts its descendants into preserve-unknown-fields
+	if p, ok := m.Extensions["x-kubernetes-preserve-unknown-fields"]; ok && p == true {
```

Comment:
this is comparing an arbitrary `interface{}` directly to a boolean? I can't remember if that exposes us to risk of runtime panics with incomparable values

Reply:
```go
func TestIncomparable(t *testing.T) {
	extensions := map[string]any{
		"trueKey": true,
		"key1":    "value",
		"key2":    2,
		"key3":    []string{"value"},
		"key4": map[string]any{
			"key": "value",
		},
		"key5": nil,
	}

	if v, ok := extensions["trueKey"]; ok && v != true {
		t.Fatal("should be equal")
	}

	if v, ok := extensions["key1"]; ok && v == true {
		t.Fatal("should not be equal")
	}
	if v, ok := extensions["key2"]; ok && v == true {
		t.Fatal("should not be equal")
	}
	if v, ok := extensions["key3"]; ok && v == true {
		t.Fatal("should not be equal")
	}
	if v, ok := extensions["key4"]; ok && v == true {
		t.Fatal("should not be equal")
	}
	if v, ok := extensions["key5"]; ok && v == true {
		t.Fatal("should not be equal")
	}
}
```

Go seems to handle it well - the above test passes when I ran on go1.19